tests/dag_builder_test.py

```diff
@@ -169,6 +155,23 @@ def test_join(self) -> None:
         self._expect_rdd(node=cities_rdd, expected_schema=schema)
 
 
+    def test_join_simple(self) -> None:
+        ab = spark.createDataFrame([], schema="struct<a: double, b: double>")
+        bc = spark.createDataFrame([], schema="struct<b: double, c: double>")
+        df = ab.join(bc, on=["b"])
```

Comment:
try testing with a `join` that has an arbitrary join condition rather than a list of columns. For example, could be

```python
ab.join(bc, on=[ab.b == bc.b])
```

Reply:
This case is a bit tricky.
If you have an arbitrary condition, PySpark does't add the "extra" project node... And if you think about it, it makes sense, because with something like: `ab.join(bc, on=[ab.b >= bc.b])`, then you could expect to have both the "b from bc" and the "b from ab" columns.
We could, at most, add a special handling for the case where it's an arbitrary condition and that condition is an "equals". Do you think it's worth it?

Reply:
> If you have an arbitrary condition, PySpark does't add the "extra" project node

exactly, that was my point. Even if you write the join I mentioned above (which compares 2 columns by equality) PySpark keeps both columns, so this `Project` node will not appear (because all columns are kept, like you mentioned). I think this projection is only added when using a list of columns in `on=[...]`

> We could, at most, add a special handling for the case where it's an arbitrary condition and that condition is an "equals". Do you think it's worth it?

I'm just worried about removing projections that should be shown (i.e. do more than just filtering a set of rows).

Reply:
Hmmm, currently we're removing that extra project in that case... I will check if there is a way to detect the difference

Reply:
Added that handling and a test